### PR TITLE
[feat] Support `columns_sorted` in row_filters with pageIndex

### LINKED ISSUE #3476: Support `columns_sorted` in row_filters
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Now
```
/// For a given set of `Column`s required for predicate `Expr` determine whether all
/// columns are sorted. Sorted columns may be queried more efficiently in the presence of
/// a PageIndex.
fn columns_sorted(_columns: &[usize], _metadata: &ParquetMetaData) -> Result<bool> {
    // TODO How do we know this?
    Ok(false)
}
```
I think we can set true with one col with pageIndex set ordered. 🤔

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
# Which issue does this PR close?

Closes #3476.

I think we can set true with only one col with pageIndex which is ordered. 🤔

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/Cargo.toml [hunk 1]
```diff
@@ -79,6 +79,7 @@ object_store = "0.5.0"
 ordered-float = "3.0"
 parking_lot = "0.12"
 parquet = { version = "22.0.0", features = ["arrow", "async"] }
+parquet-format = "4.0.0"
```
Comment (Ted-Jiang) [inline datafusion/core/Cargo.toml:82]: Should we move this logic to `arrow-rs` to avoid import this `parquet `

**Comment by Ted-Jiang** [comment]
@alamb @thinkharderdev  @tustvold  PTAL

File: datafusion/core/Cargo.toml (hunk 1 above)
Comment (thinkharderdev) [reply datafusion/core/Cargo.toml:82]: I think that makes sense

**Review by thinkharderdev — APPROVED** [review]
LGTM

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 2]
```diff
@@ -228,11 +230,32 @@ fn size_of_columns(columns: &[usize], metadata: &ParquetMetaData) -> Result<usiz
...
-    // TODO How do we know this?
+fn columns_sorted(columns: &[usize], metadata: &ParquetMetaData) -> Result<bool> {
+    // For now we only set sorted with single col with pageIndex
+    if columns.len() == 1 && metadata.page_indexes().is_some() {
+        if let Some(indexes) = metadata.page_indexes().unwrap().get(0) {
+            return match indexes.get(columns[0]).unwrap() {
+                Index::NONE
+                | Index::BOOLEAN(_)
+                | Index::BYTE_ARRAY(_)
+                | Index::FIXED_LEN_BYTE_ARRAY(_) => Ok(false),
```
Comment (liukun4515) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:241]: Do we need to support `bool, bytearray, fixed length byte array` future？

we can file a new issue to track this.

**Review by liukun4515 — APPROVED** [review]

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 2 above)
Comment (Ted-Jiang) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:241]: >Do we need to support bool, bytearray, fixed length byte array in the future？

I think not, in https://github.com/apache/arrow-datafusion/pull/3380, try put columns_sorted col ahead to filter row as much as possible, I think ` bool, bytearray, fixed length byte array` col using `page_index` 's min_max is not an `efficient` filter.

Comment (liukun4515) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:241]: Why the filter for `fixed length byte array` is not efficient?
In the parquet, spark will write decimal(20,2) and  encode the value as `fixed length byte array`, we can use the oder to accelerate the filter like `int32,int64`.

Comment (liukun4515) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:241]: I am not the expert in the parquet column ColumnOrder, but from the document mentioned above, the order is not the nature order.
The order is determined by logical type not the physical type.

File: datafusion/core/Cargo.toml (hunk 1 above)
Comment (alamb) [reply datafusion/core/Cargo.toml:82]: I think parquet `23.0.0` will have properly exposed arrow definitions (added by @tustvold  in https://github.com/apache/arrow-rs/pull/2626) 

Though now that I see this, I see the proposal is probably to move the functions `columns_sorted` and `check_is_ordered` into the `parquet` crate which makes sense to me (and will likely get reviewed by some others with more parquet knowledge)

@Ted-Jiang can you file a ticket / PR to do so?

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 3]
```diff
@@ -228,11 +230,32 @@ fn size_of_columns(columns: &[usize], metadata: &ParquetMetaData) -> Result<usiz
 /// For a given set of `Column`s required for predicate `Expr` determine whether all
 /// columns are sorted. Sorted columns may be queried more efficiently in the presence of
 /// a PageIndex.
-fn columns_sorted(_columns: &[usize], _metadata: &ParquetMetaData) -> Result<bool> {
-    // TODO How do we know this?
+fn columns_sorted(columns: &[usize], metadata: &ParquetMetaData) -> Result<bool> {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:233]: I may be misreading the spec, but the documentation on 

https://github.com/apache/parquet-format/blob/54e53e5d7794d383529dd30746378f19a12afd58/src/main/thrift/parquet.thrift#L948-L959


```

  /**
   * Stores whether both min_values and max_values are orderd and if so, in
   * which direction. This allows readers to perform binary searches in both
   * lists. Readers cannot assume that max_values[i] <= min_values[i+1], even
   * if the lists are ordered.
   */
  4: required BoundaryOrder boundary_order
```

Implies to me that `BoundaryOrder` is related to the ordering of the elements in `min_values` and `max_values` lists of the metadata and not the ordering of the actual data in the pages. Give the comments on `min_values` and `max_values` it seems like those can have elements that are not in the actual file itself


So I wonder can one conclude that the data is sorted even if `BoundaryOrder` is specified? 🤔

File: datafusion/core/Cargo.toml (hunk 1 above)
Comment (Ted-Jiang) [reply datafusion/core/Cargo.toml:82]: Sure!

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 3 above)
Comment (Ted-Jiang) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:233]: >So I wonder can one conclude that the data is sorted even if BoundaryOrder is specified? 🤔

Mostly we can assume is sorted, like in `parquet-testing`  use `parquet-tools column-index ./alltypes_tiny_pages.parquet` we can find all col with `BoundaryOrder ` are sorted.😂

As we assume `columns_sorted ` as
```
Sorted columns may be queried more efficiently in the presence of a PageIndex.
```
I think use Page-index `BoundaryOrder ` is reasonable🤔

Comment (liukun4515) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:233]: Maybe i can try to find the java code, and it shows how to generate the `boundary_order` for one ColumnIndex.

**Comment by alamb** [comment]
marking this PR as a draft pending resolution of @liukun4515 's review: https://github.com/apache/arrow-datafusion/pull/3477#discussion_r972045090

**Comment by alamb** [comment]
Closing as this PR is over a year old. Please feel free to reopen it / rebase it if you plan to keep working on it

**Pull request closed without merging**